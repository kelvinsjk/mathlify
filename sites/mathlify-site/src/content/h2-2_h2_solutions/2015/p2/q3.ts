import { Answer } from '$lib/classes/answer';
import { Expression, Polynomial } from 'mathlify';
import { specialExistence } from '$content/h2-1_h2_learn/01_functions/02_inverse/01_existence/02_practice';
import { specialInverse } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { discriminantRange } from '$content/h2-1_h2_learn/01_functions/04_more/04_discriminant/02_practice';
import svg from '$static/images/h2/fns/2015p2q3a.svg?raw';
import { QED } from '$lib/typesetting/utils';

export const answer = new Answer();

// (a)
// | 1 / 1-x^2 |
// equivalent to | ba^2 / x^2 - a^2 |
// from the learn/practice section
// where b=-1 and a=1

// (b)
// (2+x) / (1-x^2)
// equivalent to (dx+e)/(c^2 - a^2x^2)

const [a, b] = [1, -1];
const exp = new Expression([a, '/', new Polynomial([a * a, 0, -1], { ascending: true })]);
const state: Parameters<typeof specialInverse>[0] = {
	fnType: 'special',
	a,
	b,
	restriction: { type: 'right', x: a, inclusive: false },
	definition: false,
	// dummies
	c: 1,
	unknownConstants: false
};

// ai
{
	const { ans, soln } = specialExistence(state);
	const imag = `\`\`\` =html\n${svg}\n\`\`\`\n\n`;
	answer.addSubPart(ans, imag + soln + ` $${QED}$`);
}
// aii
{
	const { ans, soln } = specialInverse(state, exp);
	answer.addSubPart(ans, soln);
}

// b
const stateB: Parameters<typeof discriminantRange>[0] = {
	a: 1,
	b: 0,
	c: 1,
	d: 1,
	e: 2,
	type: 3
};
{
	const { ans, soln } = discriminantRange(stateB);
	answer.addPart(ans, soln);
}
