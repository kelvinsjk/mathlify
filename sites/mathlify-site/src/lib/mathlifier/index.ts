/**
 * generates djot string,
 * using JavaScript's template literal syntax to handle mathematical expressions (both static and dynamic)
 *
 * math: starts with ${}, terminates with \n,
 * display: starts with $${}, terminates with \n\n,
 * display env: starts with $${'align'}, etc.
 * text: starts with @${}, terminates immediately
 */
export function mathlify(strings: TemplateStringsArray, ...values: unknown[]): string {
	return mathlifyGen(modules)(strings, ...values);
}

const modules: Modules = {
	math: (x: string) => `$\`${x}\``,
	display,
	mathEnvs: {
		equation: (x: string) => display(insertEnv('equation', x)),
		align: (x: string) => display(insertEnv('align', x)),
		gather: (x: string) => display(insertEnv('gather', x)),
		alignat: (x: string) => display(insertEnv('alignat', x)),
		equationStar: (x: string) => display(insertEnv('equation*', x)),
		alignStar: (x: string) => display(insertEnv('align*', x)),
		gatherStar: (x: string) => display(insertEnv('gather*', x)),
		alignatStar: (x: string) => display(insertEnv('alignat*', x)),
	},
};

function insertEnv(env: string, content: string, args = ''): string {
	if (env === 'alignat' || env === 'alignat*') {
		let firstLine = content.split('\n')[0];
		if (!firstLine) {
			firstLine = content.split('\n')[1];
		}
		const twoN = firstLine.split('&').length;
		if (twoN % 2 !== 0)
			console.warn(`alignat should have an even number of columns, but got ${twoN} instead`);
		args = `{${Math.floor(twoN / 2)}}`;
	}
	// todo: handle this with regex
	while (content.endsWith('\n') || content.endsWith('\t') || content.endsWith(' ')) {
		content = content.slice(0, content.length - 1);
	}
	return `\\begin{${env}}${args}${content}\n\\end{${env}}`;
}

function display(x: string): string {
	return `$$\`${x}\``;
}

enum Modes {
	text, // 0
	math, // 1
	display, // 2
	mathEnv, // 3
}

export interface Modules {
	math: (x: string) => string;
	display: (x: string) => string;
	mathEnvs?: {
		equation?: (x: string) => string;
		align?: (x: string) => string;
		gather?: (x: string) => string;
		alignat?: (x: string) => string;
		equationStar?: (x: string) => string;
		alignStar?: (x: string) => string;
		gatherStar?: (x: string) => string;
		alignatStar?: (x: string) => string;
	};
}

enum MathEnvs {
	equation,
	equationStar,
	align,
	alignStar,
	gather,
	gatherStar,
	alignat,
	alignatStar,
}

const mathEnvsKeys = [
	'equation',
	'equation*',
	'align',
	'align*',
	'gather',
	'gather*',
	'alignat',
	'alignat*',
] as const;
function isMathEnv(x: string): x is (typeof mathEnvsKeys)[number] {
	return (mathEnvsKeys as readonly string[]).includes(x);
}

const envToMode: Record<(typeof mathEnvsKeys)[number], MathEnvs> = {
	equation: MathEnvs.equation,
	'equation*': MathEnvs.equationStar,
	align: MathEnvs.align,
	'align*': MathEnvs.alignStar,
	gather: MathEnvs.gather,
	'gather*': MathEnvs.gatherStar,
	alignat: MathEnvs.alignat, // triggered by $${'alignat'}{x}
	'alignat*': MathEnvs.alignatStar, // triggered by $${'alignat'}{x}
};

interface MathEnvOptions {
	mathEnv: MathEnvs;
	eqnColumns?: number;
}

// TODO: alignat cols

export function mathlifyGen(
	modules: Modules,
): (strings: TemplateStringsArray, ...values: unknown[]) => string {
	return (strings: TemplateStringsArray, ...values: unknown[]) => {
		let currentModeString = '';
		let output = '';
		let mode = Modes.text;
		let mathEnvOptions: MathEnvOptions = { mathEnv: MathEnvs.equation };
		strings.forEach((str, i) => {
			const nextVal = values[i] ?? '';
			if (mode === Modes.text) {
				if (str.endsWith('@')) {
					// continues text mode
					currentModeString += `${str.slice(0, str.length - 1)}${nextVal}`;
				} else {
					// starts new text environment
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						str,
						'',
						nextVal,
						currentModeString,
					);
				}
			} else if (mode === Modes.math) {
				// checks for \n or \r\n non-greedily
				const regex = /([^]*?)(\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end math mode
					const [, before, newline, after] = match;
					output += `${before}`;
					if (output) {
						currentModeString += modules.math(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue math mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.math(output);
					}
				}
			} else if (mode === Modes.display) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end display mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += modules.display(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue display mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.display(output);
					}
				}
			} else if (mode === Modes.mathEnv) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end mathEnv mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue mathEnv mode
					if (str.endsWith(`@`)) {
						// TODO: check this.
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
				}
			}
		});
		return currentModeString;
	};
}

// returns [Mode, acc, curr, {mathEnv, eqnColumns?}]
function startNewEnv(
	after: string,
	newline: string,
	nextVal: unknown,
	curr: string,
): [Modes, string, string, MathEnvOptions] {
	const defaultMathEnvOptions = { mathEnv: MathEnvs.equation };
	let mathEnvOptions: MathEnvOptions = defaultMathEnvOptions;
	if (after.endsWith('$')) {
		let x = `${nextVal}`;
		let mode = Modes.display;
		if (isMathEnv(x)) {
			// new math env
			mode = Modes.mathEnv;
			mathEnvOptions = { mathEnv: envToMode[x] };
			//TODO: handle alignat cols
			x = '';
		}
		// new display mode
		return [mode, x, curr + `${newline}${after.slice(0, after.length - 1)}`, mathEnvOptions];
	} else if (after.endsWith('@')) {
		// new text mode
		return [
			Modes.text,
			'',
			curr + `${newline}${after.slice(0, after.length - 1)}${nextVal}`,
			defaultMathEnvOptions,
		];
	} else {
		// new math mode
		return [Modes.math, `${nextVal}`, curr + newline + after, defaultMathEnvOptions];
	}
}

function insertMathEnv(mathEnvOptions: MathEnvOptions, acc: string, modules: Modules): string {
	const mathEnv = mathEnvOptions.mathEnv;
	if (mathEnv === MathEnvs.equation) {
		return modules.mathEnvs?.equation ? modules.mathEnvs.equation(acc) : acc;
	} else if (mathEnv === MathEnvs.equationStar) {
		return modules.mathEnvs?.equationStar ? modules.mathEnvs.equationStar(acc) : acc;
	} else if (mathEnv === MathEnvs.align) {
		return modules.mathEnvs?.align ? modules.mathEnvs.align(acc) : acc;
	} else if (mathEnv === MathEnvs.alignStar) {
		return modules.mathEnvs?.alignStar ? modules.mathEnvs.alignStar(acc) : acc;
	} else if (mathEnv === MathEnvs.gather) {
		return modules.mathEnvs?.gather ? modules.mathEnvs.gather(acc) : acc;
	} else if (mathEnv === MathEnvs.gatherStar) {
		return modules.mathEnvs?.gatherStar ? modules.mathEnvs.gatherStar(acc) : acc;
	} else if (mathEnv === MathEnvs.alignat) {
		return modules.mathEnvs?.alignat ? modules.mathEnvs.alignat(acc) : acc;
	} else if (mathEnv === MathEnvs.alignatStar) {
		return modules.mathEnvs?.alignatStar ? modules.mathEnvs.alignatStar(acc) : acc;
	}
	console.warn(`mathEnv ${mathEnv} not found`);
	return acc;
}
