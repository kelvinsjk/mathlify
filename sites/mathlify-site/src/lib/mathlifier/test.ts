import { mathlify } from './';

const str = mathlify`This is inline math: ${'x+2'}xyz${'k'}
and this is text.

This is display math: $${'x+2'}xyz
abcd
${'k'}

This is aligned:
$${'align'}
  x+2 &= ${'k'} \\\\
  y &= ${'z'}`;

console.log(str);
