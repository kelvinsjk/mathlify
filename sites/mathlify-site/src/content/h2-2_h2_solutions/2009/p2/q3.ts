import { Answer } from '$lib/classes/answer';
import { fnTerm } from 'mathlify';
import { mathlifierDj as mathlifier } from 'mathlifier';
import {
	generateAns,
	generateFracFnDefinition
} from '$content/h2-1_h2_learn/01_functions/04_more/01_self-inverse/02_practice';
import { EquationWorking, solve } from 'mathlify/working';
import { QED } from '$lib/typesetting/utils';

export const answer = new Answer();

// f(x) = ax/(bx-a)
// g(x) = 1/x

const fState: Parameters<typeof generateFracFnDefinition>[0] = {
	// (bx+k)/(cx+a)
	fnType: 'frac',
	//a: '-a',
	b: 'a',
	c: 'b',
	k: 0,
	n: 2,
	show: false
};
const fExp = generateFracFnDefinition(fState, { for: true })[1];
// a
{
	const { ans, soln } = generateAns(fState, fExp, 'f', { qed: true });
	answer.addPart(
		ans +
			mathlifier`\\
${{}}R_{f^2} = \\left( -\\infty, \\frac{a}{b} \\right) \\cup \\left( \\frac{a}{b}, \\infty \\right).`,
		soln +
			mathlifier`\n
$${{}} R_{f^2} = \\left( -\\infty, \\frac{a}{b} \\right) \\cup \\left( \\frac{a}{b}, \\infty \\right) \\; ${QED}`
	);
}
// b
{
	const soln = mathlifier`$${'align*'} R_g &= \\left( -\\infty, 0 \\right) \\cup \\left( 0, \\infty \\right)
\\\\ D_f &= \\left( -\\infty, \\frac{a}{b} \\right) \\cup \\left( \\frac{a}{b}, \\infty \\right)

Since ${'a'}
and ${'b'}
are non-zero,

$${{}} R_g \\not \\subseteq D_f

so the composite function ${'fg'}
does not exist ${QED}`;
	answer.addPart(
		mathlifier`The composite function ${'fg'}
does not exist as ${{}} {R_g \\not \\subseteq D_f.}`,
		soln
	);
}
// c
{
	const working = new EquationWorking(fnTerm('f^{-1}', 'x'), 'x');
	working.subIn({ 'f^{-1}': fExp });
	working.crossMultiply({ hide: true }).expand().swapSides({ hide: true });
	working.makeRhsZero().factorize.commonFactor();
	const { roots, working: rootsWorking } = solve.zeroProduct(working.eqn, 'x', { qed: true });
	const soln = mathlifier`$${'gather*'} ${working}

$${'align*'} ${rootsWorking}	`;
	const ans = mathlifier`${'x'} = ${roots.join(' \\text{ or }')}.`;
	answer.addPart(ans, soln);
}
