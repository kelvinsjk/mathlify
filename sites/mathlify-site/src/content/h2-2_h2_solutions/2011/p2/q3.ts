import { Answer } from '$lib/classes/answer';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Polynomial, sum } from 'mathlify';
import { logInverse } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { inverseRelationshipSolver } from '$content/h2-1_h2_learn/01_functions/02_inverse/05_relationship/02_practice-1';
import svg from '$static/images/h2/fns/2011p2q3b.svg?raw';
import { logTerm } from 'mathlify/fns';

export const answer = new Answer();

// ! Complete
// ln(2x+1) + 3

const argument = new Polynomial([2, 1]);
const b = 3;
const yMinusB = sum('y', -b);
const f = sum(logTerm(argument), b);

// a
{
	const { ans, soln } = logInverse(f, argument, yMinusB);
	answer.addPart(ans, soln);
}
// b
{
	const soln = `\`\`\` =html\n${svg}\n\`\`\`\n\n`;
	const ans = mathlifier`[Sketch](#soln-2).`;
	answer.addPart(ans, soln);
}
// c
const state = {
	fnType: 'log',
	a: 1,
	c: 2,
	b: 3
} as const;
{
	const { ans, soln } = inverseRelationshipSolver(state, f, {
		precision: 4,
		tys2011ExtraStep: true
	});
	answer.addPart(ans, soln);
}
