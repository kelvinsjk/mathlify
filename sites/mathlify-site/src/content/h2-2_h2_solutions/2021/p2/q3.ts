import { Answer } from '$lib/classes/answer';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Expression, fnTerm, Polynomial, quotient } from 'mathlify';
import { fractionalInverseWorking } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { EquationWorking, ExpressionWorking } from 'mathlify/working';
import { QED } from '$lib/typesetting/utils';

export const answer = new Answer();

// ! Complete, but can be further abstracted

// 1/2 x^2 + 3
// (x+1)/(5x-1)
// (x+a)(2x+b)

const h = new Polynomial([[1, '/', 2], 0, 3]);
const g = quotient(new Polynomial([1, 1]), new Polynomial([5, -1]));
const verbatim = true;
const num = new Polynomial([1, 'a']);
const den = new Polynomial([2, 'b']);
const f = quotient(num, den);

// ai
{
	const x = 2;
	const h2 = fnTerm('h', x);
	const gh2 = fnTerm('g', h2);
	const working = new ExpressionWorking(gh2);
	working.subIn({ h }, { verbatim }).simplify().subIn({ g }, { verbatim }).simplify();
	const finalAns = working.expression;

	const soln = mathlifier`$${'align*'} ${working}`;
	const ans = mathlifier`${{}} ${gh2} = ${finalAns}.`;
	answer.addSubPart(ans, soln);
}
// aii
{
	const b = quotient(14, 10);
	const working = new EquationWorking(g, b);
	working.crossMultiply().expand().solve.linear('x');
	const x = working.eqn.rhs;
	const soln = mathlifier`$${'gather*'} g(x) = ${b.valueOf()} \\\\ ${working}`;
	const ans = mathlifier`${{}} x = ${x}.`;
	answer.addSubPart(ans, soln);
}
answer.newPart();
// bi
{
	const working = new EquationWorking(den, 0);
	const { root: k } = working.solve.linear('x');
	const soln = mathlifier`Note that division by zero is undefined
$${'gather*'}${working}

Hence ${{}}k=${k}
and it must be excluded from the domain of ${'f'}
since ${'f'}
is not defined when ${{}} x = ${k}`;
	const ans = mathlifier`${{}} k = ${k}.
\\
It must be excluded as ${'f'}
is not defined when ${{}} {x = ${k}.}`;
	answer.addSubPart(ans, soln);
}
// bii
let fInv: Expression;
{
	const { working, exp: fInv1 } = fractionalInverseWorking(f, false, { reportInverse: true });
	const b = -1;
	const a = quotient(-1, 2);
	fInv = f.subIn({ b });
	const soln =
		working +
		mathlifier`$${'align*'} f(x) &= f^{-1}(x) \\\\ ${f} &= ${fInv1} 

Since this is valid for all ${'x'}
in the domain of ${'f'},
by comparing,

$${{}} b = ${b} ${QED}

Substituting this into ${'f(x)'}:

$${{}} f(x) = ${fInv}

We observe that if ${{}} {a = ${a},}
then ${{}} {f(x) = ${fInv.subIn({ a })} = \\frac{1}{2}}
which is not a one-to-one function and does not have an inverse.

Hence ${'a'}
can take all real values except ${a}

$${{}} a \\in \\mathbb{R}, a \\neq ${a} ${QED}`;
	const ans = mathlifier`${{}} b = ${b}.
\\
${{}} a \\in \\mathbb{R}, a \\neq ${a}.`;
	answer.addSubPart(ans, soln);
}
// bIII
{
	const x = -4;
	const fInvX = fnTerm('f^{-1}', x);
	const working = new ExpressionWorking(fInvX, { startOnFirstLine: true });
	working
		.subIn({ 'f^{-1}': fnTerm('f', x) })
		.subIn({ f: fInv }, { verbatim })
		.simplify()
		.expandNegativeIntoQuotient();
	const finalAns = working.expression;
	const ans = mathlifier`${{}} f^{-1}(${x}) = ${finalAns}.`;
	const soln = mathlifier`$${'align*'} ${working}`;
	answer.addSubPart(ans, soln);
}
