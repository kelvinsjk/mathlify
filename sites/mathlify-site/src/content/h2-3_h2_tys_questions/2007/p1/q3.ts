/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Expression, i, sum } from 'mathlify';
import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

// ww^* + 2w = 3 + 4i

const w = new Expression('w');
const wStar = new Expression('w^*');
const lhs = sum([w, wStar], [2, w]);
const rhs = sum(3, [4, i]);
const aPlusBI = sum('a', [i, 'b']);

// a
{
	question.addPart(mathlifier`Out of syllabus (complex locus).`, { marks: 3 });
}
// b
{
	question.addPart(
		mathlifier`The complex number ${w}
is such that

$${lhs} = ${rhs},

where ${wStar}
is the complex conjugate of ${w}.
Find ${w}
in the form ${aPlusBI},
where ${'a'}
and ${'b'}
are real.`,
		{ marks: 4 }
	);
}
