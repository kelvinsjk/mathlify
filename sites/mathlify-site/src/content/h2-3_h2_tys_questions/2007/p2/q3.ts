/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Polynomial } from 'mathlify';
import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

const ascending = true;
const poly1 = new Polynomial([1, 1], { ascending });
const poly2 = new Polynomial([4, -1], { ascending });
const poly3 = new Polynomial([1, 0, 2], { ascending });

const exp1 = `{(${poly1})^n,}`;
const exp2 = `{(${poly2})^{\\frac{3}{2}} (${poly3})^{\\frac{3}{2}}}`;

// a
question.addPart(
	mathlifier`By successively differentiating ${exp1}
find Maclaurin's series for ${exp1}
up to and including the term in ${'x^3'}.`,
	{ marks: 4 }
);
// b
question.addPart(
	mathlifier`Obtain the expansion of ${exp2}
up to and including the term in ${'x^3'}.`,
	{ marks: 5 }
);
// c
question.addPart(
	mathlifier`Find the set of values of ${'x'}
for which the expansion in part (b) is valid.`,
	{ marks: 2 }
);
