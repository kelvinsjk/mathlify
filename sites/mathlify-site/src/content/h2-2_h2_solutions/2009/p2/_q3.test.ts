import { expect, it } from 'vitest';
import { answer } from './q3';
import { mathlifierDj as mathlifier } from 'mathlifier';

const ans = answer.answer;
const actual: typeof ans = {
	parts: [
		{
			body:
				mathlifier`${{}} f^{-1}(x) = \\frac{ax}{bx - a}.` +
				'\\\n' +
				mathlifier`${{}}f^{2}(x) = x.` +
				'\\\n' +
				mathlifier`${{}}R_{f^2} = \\left( -\\infty, \\frac{a}{b} \\right) \\cup \\left( \\frac{a}{b}, \\infty \\right).`
		},
		{
			body: mathlifier`The composite function ${'fg'}
does not exist as ${{}} {R_g \\not \\subseteq D_f.}`
		},
		{
			body: mathlifier`${{}}x = 0 \\text{ or }\\frac{2a}{b}.`
		}
	]
};

it('2009p2q3', () => {
	expect(answer).toMatchSnapshot();
	expect(ans).toEqual(actual);
});
