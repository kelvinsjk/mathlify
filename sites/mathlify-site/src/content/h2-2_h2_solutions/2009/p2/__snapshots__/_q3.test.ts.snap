// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2009p2q3 1`] = `
Answer {
  "answer": {
    "parts": [
      {
        "body": "$\`  f^{-1}(x) = \\frac{ax}{bx - a}. \`\\
$\` f^{2}(x) = x. \`\\
$\` R_{f^2} = \\left( -\\infty, \\frac{a}{b} \\right) \\cup \\left( \\frac{a}{b}, \\infty \\right). \`",
      },
      {
        "body": "The composite function $\` fg \`
does not exist as $\`  {R_g \\not \\subseteq D_f.} \`",
      },
      {
        "body": "$\` x = 0 \\text{ or }\\frac{2a}{b}. \`",
      },
    ],
  },
  "solution": {
    "parts": [
      {
        "body": "$$\` \\begin{gather*} \\text{Let }  y = \\frac{ax}{bx - a}
	\\\\ y\\left( bx - a \\right) = ax
	\\\\ ybx - ya = ax
	\\\\ ybx - ax = ya
	\\\\ x\\left( yb - a \\right) = ya
	\\\\ x = \\frac{ya}{yb - a} \\\\ f^{-1}(x) = \\frac{xa}{xb - a} \\; \\blacksquare
\\end{gather*} \`$$\` \\begin{align*} f^{2}(x) &= ff(x) \\\\ &= ff^{-1}(x) \\\\ &= x \\; \\QED
\\end{align*} \`

$$\`  R_{f^2} = \\left( -\\infty, \\frac{a}{b} \\right) \\cup \\left( \\frac{a}{b}, \\infty \\right) \\; \\QED  \`",
      },
      {
        "body": "$$\` \\begin{align*} R_g &= \\left( -\\infty, 0 \\right) \\cup \\left( 0, \\infty \\right)
\\\\ D_f &= \\left( -\\infty, \\frac{a}{b} \\right) \\cup \\left( \\frac{a}{b}, \\infty \\right)
\\end{align*} \`

Since $\` a \`
and $\` b \`
are non-zero,

$$\`  R_g \\not \\subseteq D_f \`

so the composite function $\` fg \`
does not exist $\` \\QED  \`",
      },
      {
        "body": "$$\` \\begin{gather*} f^{-1}\\left(x\\right) = x
	\\\\ \\frac{ax}{bx - a} = x
	\\\\ ax = x^2b - xa
	\\\\ x^2b - 2xa = 0
	\\\\ x\\left( xb - 2a \\right) = 0
\\end{gather*} \`

$$\` \\begin{align*} x &=0\\quad\\blacksquare  &&\\quad\\text{ or }\\quad& xb - 2a &=0
	\\\\ & &&\\quad& xb &=2a
	\\\\ & &&\\quad& x &=\\frac{2a}{b}\\quad\\blacksquare
\\end{align*} \`",
      },
    ],
  },
}
`;
